### ISSUE TITLE
avoid exception in logbytecounter

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10999]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| andrea reported|
|-|-|
|Trac ID|trac#1437|
|Type|defect|
|Created|2006-01-22 23:16:27Z|

Attachments:

* [web2-log.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/637/6371a5630b9dc4ebd3cd2b721357f87eee6c5f0c/cc564cd085872599043144e86419cc3f64746473.diff) (903 bytes) - added by jerub on 2006-02-26 02:20:24Z - don't wrap a None response.stream. (includes test)

<details><summary>Searchable metadata</summary>

```
trac-id__1437 1437
type__defect defect
reporter__andrea andrea
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1137971787000000 1137971787000000
changetime__1142110410000000 1142110410000000
version__ 
owner__ 
cc__jknight cc__andrea cc__jerub
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
http://www.cpushare.com/hg/Twisted/?cs=57753be7d975
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
sorry, updated link after rebasing over cpushare.com/hg/TwistedSVN  
  
http://www.cpushare.com/hg/Twisted?cs=4f37b5b687ee
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

Reviewing the traceback and patch from the link, I don't think this is the best way of resolving the ticket. The _LogByteCounter object is a wrapper around a file-like-object, but in this case seems to have been passed 'None' instead of a file-like-object.

```python
def logFilter(request, response, startTime=None):
# ...
    response.stream=_LogByteCounter(response.stream, _log)
    return response
```

If response.stream is None, attaching that filter shouldn't occur. I don't know how response.stream can be None, so maybe this is exposing a deeper bug. It would be nice to have more information about how this could occur.

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

I also had some doubt about this being a band aid and not a real fix. However it seems the response stream can be NULL sometime in the filter chain call, see http:writeResponse:

        if response.code != responsecode.NOT_MODIFIED:
            # Not modified response is *special* and doesn't get a content-length.
            if response.stream is None:
                response.headers.setHeader('content-length', 0)

If it can be null there, it can be null as well when the logFilter is being invoked, hence my patch seems valid to me and not just a band-aid.

Clearly a change like the below would look less like a band-aid but then I'm afraid we'd stop seeing those request in the logs so it would be worse.

    if response.stream is not None:
          response.stream=_LogByteCounter(response.stream, _log)

I'm unsure what has triggered the bug in the first place, it was random internet traffic. Now the whole cpushare site (not only klive) is running on top of Cheetah + web2 and I never had a single exception since I deployed it, so at least in practice it works.

Clearly I welcome any suggestion for a cleaner fix, but I'm quite confortable that this fix is better than nothing and it's not (necessarily ;) hiding a bug.

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

hmm, looks like I need to learn trac wiki formatting... or to press preview sometime ;)

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)|@jerub commented|
|-|-|

I see the need for a better solution, along the lines of what you suggest. Otherwise it looks like requests won't get logged if there isn't a response.stream.

Can you prepare a patch?

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

Yes, stream is allowed to be None, and the logger should handle it.

**Comment by twisted-trac** [comment]
|<img alt="foom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|foom set status to `closed`|
|-|-|

(In [16246]) Check if the stream is None before trying to wrap it.
Fixes [#10999](https://github.com/twisted/twisted/issues/10999).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
